Rust error when trying to access an element of an array within a ROW (array, array)

The following SQL passes:

```
CREATE TABLE row_of_arr_tbl(
c2 ROW(i2 INT ARRAY NULL, v2 VARCHAR ARRAY NOT NULL));


CREATE MATERIALIZED VIEW row_of_arr_elmnt_access AS SELECT
row_of_arr_tbl.c2.v2[2],
row_of_arr_tbl.c2[2][2]
FROM row_of_arr_tbl;
```
Given the following input:
```
{
    "c2": {"i2": [1, 2, None], "v2": ["bye", "see you later!"]}
}
```

But produces the following error:

> **error[E0308]: `match` arms have incompatible types**
>    --> projects/project-b08435ce4cef84a9e68a80fb95aa25a8dd888f825015134a50b849868807170d/src/main.rs:225:32
>     |
> 223 |                   indexNN_(match &(*t_3).2 {
>     |  __________________________-
> 224 | |                     None => &None,
>     | |                             ----- this is found to be of type `&std::option::Option<Arc<Vec<std::option::Option<_>>>>`
> 225 | |                     Some(x) => &x.1
>     | |                                ^^^^ expected `&Option<Arc<Vec<Option<_>>>>`, found `&Arc<Vec<Option<SqlString>>>`
> 226 | |                 }, 1isize).clone(), 
>     | |_________________- `match` arms have incompatible types
>     |
>     = note: expected reference `&std::option::Option<Arc<Vec<std::option::Option<_>>>>`
>                found reference `&Arc<Vec<std::option::Option<feldera_sqllib::SqlString>>>`
> 
> **error[E0308]: `match` arms have incompatible types**
>    --> projects/project-b08435ce4cef84a9e68a80fb95aa25a8dd888f825015134a50b849868807170d/src/main.rs:229:32
>     |
> 227 |                   indexNN_(match &(*t_3).2.as_ref() {
>     |  __________________________-
> 228 | |                     None => &None,
>     | |                             ----- this is found to be of type `&std::option::Option<Arc<Vec<std::option::Option<_>>>>`
> 229 | |                     Some(x) => &x.1
>     | |                                ^^^^ expected `&Option<Arc<Vec<Option<_>>>>`, found `&Arc<Vec<Option<SqlString>>>`
> 230 | |                 }, 1isize).clone())
>     | |_________________- `match` arms have incompatible types
>     |
>     = note: expected reference `&std::option::Option<Arc<Vec<std::option::Option<_>>>>`
>                found reference `&Arc<Vec<std::option::Option<feldera_sqllib::SqlString>>>`

--- a/crates/adapters/src/util.rs
+++ b/crates/adapters/src/util.rs
@@ -1,17 +1,24 @@
 use std::borrow::Cow;
 use std::path::Path;
 use std::{
-    error::Error,
     fs::File,
-    future::Future,
     io::{Error as IoError, Write},
+};
+
+#[cfg(feature = "with-deltalake")]
+use std::{
+    error::Error,
+    future::Future,
     pin::Pin,
 };
 
+#[cfg(feature = "with-deltalake")]
 use futures::channel::oneshot;
 use tempfile::NamedTempFile;
+#[cfg(feature = "with-deltalake")]
 use tokio::{spawn, task::JoinHandle};
 
+#[cfg(feature = "with-deltalake")]
 pub(crate) fn root_cause(mut err: &dyn Error) -> &dyn Error {
     while let Some(source) = err.source() {
         err = source;
@@ -100,6 +107,7 @@ producer  ─┬─►│ │ │ │ │ │ ├─┼─►│worker2├──
   sends a special Sync message to the completions queue. Upon receiving
   this message, the consumer sends an acknowledgement to the sync channel.
 */
+#[cfg(feature = "with-deltalake")]
 pub(crate) struct JobQueue<I, O> {
     /// The producer side of the job queue.
     job_sender: async_channel::Sender<(I, oneshot::Sender<O>)>,
@@ -120,11 +128,13 @@ pub(crate) struct JobQueue<I, O> {
 // Every message in the completions channel contains the receiving side of the
 // oneshot channel that will contain the result of the completed job, or a special
 // Sync message.
+#[cfg(feature = "with-deltalake")]
 enum Completion<O> {
     Completion(oneshot::Receiver<O>),
     Sync,
 }
 
+#[cfg(feature = "with-deltalake")]
 impl<I, O> JobQueue<I, O>
 where
     I: Send + 'static,
@@ -222,6 +232,7 @@ where
     }
 }
 
+#[cfg(feature = "with-deltalake")]
 impl<I, O> Drop for JobQueue<I, O> {
     fn drop(&mut self) {
         self.consumer.abort();
@@ -232,6 +243,7 @@ impl<I, O> Drop for JobQueue<I, O> {
 }
 
 #[cfg(test)]
+#[cfg(feature = "with-deltalake")]
 mod test {
     use std::sync::{Arc, Mutex};
 

--- a/crates/adapters/src/util.rs
+++ b/crates/adapters/src/util.rs
@@ -6,11 +6,7 @@ use std::{
 };
 
 #[cfg(feature = "with-deltalake")]
-use std::{
-    error::Error,
-    future::Future,
-    pin::Pin,
-};
+use std::{error::Error, future::Future, pin::Pin};
 
 #[cfg(feature = "with-deltalake")]
 use futures::channel::oneshot;

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
@@ -31,7 +31,6 @@ import org.dbsp.sqlCompiler.compiler.errors.SourcePositionRange;
 import org.dbsp.sqlCompiler.compiler.errors.UnimplementedException;
 import org.dbsp.sqlCompiler.compiler.errors.UnsupportedException;
 import org.dbsp.sqlCompiler.compiler.frontend.ExpressionCompiler;
-import org.dbsp.sqlCompiler.compiler.frontend.calciteObject.CalciteObject;
 import org.dbsp.sqlCompiler.compiler.visitors.VisitDecision;
 import org.dbsp.sqlCompiler.compiler.visitors.inner.InnerVisitor;
 import org.dbsp.sqlCompiler.compiler.visitors.inner.Simplify;
@@ -1067,25 +1066,52 @@ public class ToRustInnerVisitor extends InnerVisitor {
             case SQL_INDEX: {
                 DBSPType collectionType = expression.left.getType();
                 DBSPType indexType = expression.right.getType();
-                DBSPTypeArray vec = collectionType.to(DBSPTypeArray.class);
+                DBSPTypeArray arrayType = collectionType.to(DBSPTypeArray.class);
                 this.builder.append("index")
                         .append(expression.left.getType().nullableUnderlineSuffix())
-                        .append(vec.getElementType().nullableUnderlineSuffix())
+                        .append(arrayType.getElementType().nullableUnderlineSuffix())
                         .append(indexType.nullableUnderlineSuffix())
                         .append("(");
-                // Special case for Tuple<Option<Vec<>>>
                 boolean leftDone = false;
+                // Cases:
+                // - t: Tuple<Array<>>: &t.0
+                // - t: Tuple<Option<Array<>>>: &t.0
+                // - t: Option<Tuple<Array<>>>: &match t { None => None, Some(x) => Some(x.0.clone()) }
+                // - t: Option<Tuple<Option<Array>>>>: &match t { None => None, Some(x) => x.0.clone() }
                 if (expression.left.is(DBSPFieldExpression.class) && collectionType.mayBeNull) {
+                    // One of last 3 cases
                     DBSPFieldExpression field = expression.left.to(DBSPFieldExpression.class);
-                    if (field.expression.is(DBSPFieldExpression.class)) {
-                        this.builder.append("match &");
-                        field.expression.accept(this);
+                    DBSPTypeTupleBase tupleType = field.expression
+                            .getType()
+                            .to(DBSPTypeTupleBase.class);
+                    boolean tupleMaybeNull = tupleType.mayBeNull;
+                    boolean arrayMayBeNull = tupleType
+                            .getFieldType(field.fieldNo).mayBeNull;
+                    if (tupleMaybeNull) {
+                        // one of last 2 cases
                         leftDone = true;
-                        this.builder.append(" {").increase()
-                                .append("None => &None,").newline()
-                                .append("Some(x) => &x.")
-                                .append(field.fieldNo).newline()
-                                .decrease().append("}");
+                        if (arrayMayBeNull) {
+                            this.builder.append("&match &");
+                            field.expression.accept(this);
+                            this.builder.append(" {").increase()
+                                    .append("None => None,").newline()
+                                    .append("Some(x) => x.")
+                                    .append(field.fieldNo);
+                            this.builder.append(".clone()")
+                                    .newline()
+                                    .decrease().append("}");
+                        } else {
+                            this.builder.append("&match &");
+                            field.expression.accept(this);
+                            this.builder.append(" {").increase()
+                                    .append("None => None,").newline()
+                                    .append("Some(x) => ");
+                            this.builder.append("Some(x.")
+                                    .append(field.fieldNo)
+                                    .append(".clone())")
+                                    .newline()
+                                    .decrease().append("}");
+                        }
                     }
                 }
                 if (!leftDone) {

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/StructTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/StructTests.java
@@ -137,10 +137,22 @@ public class StructTests extends SqlIoTest {
     }
 
     @Test
+    public void issue3671() {
+        this.getCCS("""
+                CREATE TABLE row_of_arr_tbl(
+                c2 ROW(i2 INT ARRAY NULL, v2 VARCHAR ARRAY NOT NULL));
+                
+                CREATE VIEW v AS SELECT
+                row_of_arr_tbl.c2.i2[2],
+                row_of_arr_tbl.c2.v2[2],
+                row_of_arr_tbl.c2[2][2]
+                FROM row_of_arr_tbl;""");
+    }
+
+    @Test
     public void structArrayTest() {
         String ddl = """
-            CREATE TYPE address_typ AS (
-               street          VARCHAR ARRAY);
+            CREATE TYPE address_typ AS (street VARCHAR ARRAY);
             CREATE TABLE PERS(p0 address_typ);
             CREATE VIEW V AS SELECT PERS.p0.street[1] FROM PERS;
             """;

